cdap-common/src/main/java/io/cdap/cdap/common/leveldb/FileMetaDataRewriter.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright © 2022 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.common.leveldb;
+
+import io.cdap.cdap.common.lang.ClassRewriter;
+import org.objectweb.asm.ClassReader;
+import org.objectweb.asm.ClassVisitor;
+import org.objectweb.asm.ClassWriter;
+import org.objectweb.asm.MethodVisitor;
+import org.objectweb.asm.Opcodes;
+import org.objectweb.asm.Type;
+import org.objectweb.asm.commons.AdviceAdapter;
+import org.objectweb.asm.commons.Method;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.concurrent.Executor;
+import javax.annotation.Nullable;
+
+public class FileMetaDataRewriter implements ClassRewriter {
+  private static final String FILE_META_DATA_CLASS_NAME = "org.iq80.leveldb.impl.FileMetaData";
+  private static final String FILE_META_DATA_UTIL_CLASS_RESOURCE = "io/cdap/cdap/common/leveldb/FileMetaDataUtil";
+  private static final String NORMALIZE_METHOD_NAME = "normalize";
+  private static final Type INTERNAL_KEY_TYPE = Type.getObjectType("org/iq80/leveldb/impl/InternalKey");
+
+  public boolean needRewrite(String className) {
+    return FILE_META_DATA_CLASS_NAME.equals(className);
+  }
+
+  @Override
+  public byte[] rewriteClass(String className, InputStream input) throws IOException {
+    ClassReader cr = new ClassReader(input);
+    ClassWriter cw = new ClassWriter(ClassWriter.COMPUTE_FRAMES);
+
+    Method method = new Method("<init>", Type.VOID_TYPE, new Type[] {
+      Type.LONG_TYPE, Type.LONG_TYPE, INTERNAL_KEY_TYPE, INTERNAL_KEY_TYPE});
+    cr.accept(new ClassVisitor(Opcodes.ASM7, cw) {
+      @Override
+      public MethodVisitor visitMethod(int access, String name, String descriptor, String signature, String[] exceptions) {
+        MethodVisitor mv = super.visitMethod(access, name, descriptor, signature, exceptions);
+        if (!method.equals(new Method(name, descriptor))) {
+          return mv;
+        }
+        return new AdviceAdapter(Opcodes.ASM7, mv, access, name, descriptor) {
+          @Override
+          protected void onMethodEnter() {
+            // Each long parameter takes 2 slots: 1,2 for number and 3,4 for filesize
+            // So InternalKey parameters are vars 5 and 6
+            mv.visitVarInsn(Opcodes.ALOAD, 5);
```

Comment:
It's better to use methods in `GeneratorAdapter` to make it easier to read (e.g. `loadLocal`, `invokeStatic`, etc).

Reply:
Done